includes/class-cron-options-cpt.php

```diff
@@ -0,0 +1,375 @@
+		}
+
+		return empty( $exists ) ? false : array_shift( $exists );
+	}
+
+	/**
+	 * Create a job post, respecting whether or not Core is ready for CPTs
+	 */
+	private function create_job( $job_post ) {
+		// Limit how many events to insert at once
+		if ( ! Lock::check_lock( self::LOCK, 5 ) ) {
+			return false;
+		}
+
+		// If called before `init`, we need to insert directly because post types aren't registered earlier
+		if ( did_action( 'init' ) ) {
+			wp_insert_post( $job_post );
+		} else {
+			global $wpdb;
+
+			// Additional data needed to manually create a post
+			$job_post = wp_parse_args( $job_post, array(
+				'post_author'       => 0,
+				'comment_status'    => 'closed',
+				'ping_status'       => 'closed',
+				'post_parent'       => 0,
+				'post_modified'     => current_time( 'mysql' ),
+				'post_modified_gmt' => current_time( 'mysql', true ),
+			) );
+
+			// Some sanitization in place of `sanitize_post()`, which we can't use this early
+			foreach ( array( 'post_title', 'post_name', 'post_content_filtered' ) as $field ) {
+				$job_post[ $field ] = sanitize_text_field( $job_post[ $field ] );
+			}
+
+			// Duplicate some processing performed in `wp_insert_post()`
+			$charset = $wpdb->get_col_charset( $wpdb->posts, 'post_title' );
+			if ( 'utf8' === $charset ) {
+				$job_post['post_title'] = wp_encode_emoji( $job_post['post_title'] );
+			}
+
+			$job_post = wp_unslash( $job_post );
+
+			// Set this so it isn't empty, even though it serves us no purpose
+			$job_post['guid'] = esc_url( add_query_arg( self::POST_TYPE, $job_post['post_name'], home_url( '/' ) ) );
+
+			// Create the post
+			$inserted = $wpdb->insert( $wpdb->posts, $job_post );
+
+			// Clear caches for new posts once the post type is registered
+			if ( $inserted ) {
+				$this->posts_to_clean[] = $wpdb->insert_id;
+			}
+		}
+
+		// Allow more events to be created
+		Lock::free_lock( self::LOCK );
+	}
+
+	/**
+	 * Remove an event's CPT entry
+	 *
+	 * @param $timestamp  int     Unix timestamp
+	 * @param $action     string  name of action used when the event is registered (unhashed)
+	 * @param $instance   string  md5 hash of the event's arguments array, which Core uses to index the `cron` option
+	 *
+	 * @return bool
+	 */
+	private function delete_job( $timestamp, $action, $instance ) {
+		$job = $this->job_exists( array(
+			'name'             => $this->event_name( $timestamp, $action, $instance ),
+			'post_type'        => self::POST_TYPE,
+			'post_status'      => self::POST_STATUS,
+			'suppress_filters' => false,
+			'posts_per_page'   => 1,
+		) );
+
+		if ( ! $job ) {
+			return false;
+		}
+
+		// If called before `init`, we need to delete directly because post types aren't registered earlier
+		if ( did_action( 'init' ) ) {
+			wp_delete_post( $job->ID, true );
+		} else {
+			global $wpdb;
+
+			$wpdb->delete( $wpdb->postmeta, array( 'post_id' => $job->ID, ) );
+			$wpdb->delete( $wpdb->posts, array( 'ID' => $job->ID, ) );
```

Comment:
What if instead of deleting during the process (which may be slow), we flag events as completed (maybe post_stutus => private) and filter them them when getting the posts? And then a separate job goes through and deletes them.